### ISSUE TITLE
CA1508 false positive

### ISSUE THREAD

**Issue by crfrolik** [dotnet/roslyn-analyzers#6520]
### Analyzer

**Diagnostic ID**: 
CA1508: Avoid dead conditional code
https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508

### Analyzer source

`dotnet --version` prints `6.0.406`

_OR_

```
"Microsoft.CodeAnalysis.NetAnalyzers" Version="7.0.0"
```

### Steps To Reproduce

See code snippet at bottom.

### Expected behavior

No error.

### Actual behavior

CA1508 is generated on lines 55, 56, and 57.

## Additional context

```csharp
using System

public class Parser
{
  private static ParsedPredicate CreateRandomPredicate()
  {
    var primitives = new[] {
      PredicateTypePrimitive.Identifier,
      PredicateTypePrimitive.Boolean,
      PredicateTypePrimitive.Integer,
      PredicateTypePrimitive.String,
    };

    var rnd = new Random();
    var p = new ParsedPredicate();
    p.Left.TypePrimitive = primitives[rnd.Next(0, 4)];
    p.Right.TypePrimitive = primitives[rnd.Next(0, 4)];
    return p;
  }

  private static ParsedPredicate ParsePredicate()
  {
    // Parse the predicate expression.
    var predicate = CreateRandomPredicate();

    // The parser returns null for an empty expression.
    if (predicate is null)
    {
      return null;
    }

    // One of the operands must be an identifier, and the
    // other must be a literal value.
    PredicateOperand identifier;
    PredicateOperand literal;

    if (predicate.Left.TypePrimitive == PredicateTypePrimitive.Identifier)
    {
      identifier = predicate.Left;
      literal = predicate.Right;
    }
    else
    {
      identifier = predicate.Right;
      literal = predicate.Left;
    }

    // Verify that the identifier is a valid identifier
    if (identifier.TypePrimitive != PredicateTypePrimitive.Identifier)
    {
      throw new InvalidOperationException();
    }

    // Verify that the other operand is a literal value.
    if (literal.TypePrimitive != PredicateTypePrimitive.Boolean &&
        literal.TypePrimitive != PredicateTypePrimitive.Integer &&
        literal.TypePrimitive != PredicateTypePrimitive.String)
    {
      throw new InvalidOperationException();
    }

    return predicate;
  }
}

public class PredicateOperand
{
  public PredicateTypePrimitive TypePrimitive { get; set; } = PredicateTypePrimitive.String;
}

public enum PredicateTypePrimitive
{
  Boolean = 400,
  Integer = 401,
  Identifier = 402,
  String = 403,
}

public class ParsedPredicate
{
  public PredicateOperand Left { get; } = new ();
  public PredicateOperand Right { get; } = new ();
}
```
